test/TestApps/PushNotificationsTestApp/main.cpp

```diff
@@ -0,0 +1,388 @@
+    switchMapping["UnregisterActivator"] = UnitTest::unregisterActivator;
+}
+
+bool ChannelRequestUsingNullRemoteId()
+{
+    winrt::hresult hr = S_OK;
+
+    try
+    {
+        auto channelOperation = PushNotificationManager::CreateChannelAsync(GUID_NULL).get();
+    }
+    catch (...)
+    {
+        auto channelRequestException = hresult_error(to_hresult());
+        hr = channelRequestException.code();
+    }
+
+    return (hr == E_INVALIDARG);
+}
+
+bool ChannelRequestUsingRemoteId()
+{
+    wil::unique_handle channelEvent = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+    auto hr = S_OK;
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(remoteId1);
+
+    channelOperation.Progress(
+        [](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& /* sender */,
+            PushNotificationCreateChannelStatus const& args)
+        {
+            if (args.status == PushNotificationChannelStatus::InProgress)
+            {
+            }
+            else if (args.status == PushNotificationChannelStatus::InProgressRetry)
+            {
+            }
+        });
+
+    // Setup the completed event handler
+    channelOperation.Completed(
+        [&channelEvent](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            AsyncStatus const /* asyncStatus */)
+        {
+            auto result = sender.GetResults();
+            if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
+            {
+                auto channelUri = result.Channel().Uri();
+                auto channelExpiry = result.Channel().ExpirationTime();
+            }
+            else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
+            {
+            }
+
+            SetEvent(channelEvent.get());
+        });
+
+    // The maximum amount of time it takes for channel request to be obtained - 16mins
+    if (WAIT_OBJECT_0 != WaitForSingleObject(channelEvent.get(), 960000 /* milliseconds */))
+    {
+        channelOperation.Cancel();
+        hr = ERROR_TIMEOUT;
+    }
+    else
+    {
+        channelOperation.Close(); // Do not call getresults after this
+    }
+
+    return (hr == S_OK);
+}
+
+bool MultipleChannelRequestUsingSameRemoteId()
+{
+    wil::unique_handle channelEvent = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+    auto hr = S_OK;
+
+    auto channelOperation1 = PushNotificationManager::CreateChannelAsync(remoteId1);
+    auto channelOperation2 = PushNotificationManager::CreateChannelAsync(remoteId1);
+
+    channelOperation2.Completed(
+        [&channelEvent, &hr](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            AsyncStatus const /* asyncStatus */)
+        {
```

Comment:
How are we sure that we the HRESULT from the second channel request is going to be WPN_E_OUTSTANDING_CHANNEL_REQUEST? Is it because the first channel request will take long enough?

Reply:
What if the first channel request returns a cached channel (thinking on a request that is processed really quick)? How are we preventing this scenario in the tests?

Reply:
Let me take you to the design back again:

The sequence of events happen in the order as soon as you call CreateChannelAsync:

1. Guid null check
2. check for a remoteId that is currently being serviced, if none, then we set the s_remoteIdInProgress = true; for the first request.
3. co_await - Resume_background - Allows the second request to come in
4. while the first one is being serviced, the second one is checking for remoteIdInProgress, as it is true it immediately returns WPN_E_OUTSTANDING_CHANNEL_REQUEST. The second one would be pretty quick operation in comparison to the first one as the first one goes across the COM no matter if the channel is cached. So far I have had success in this test case passing 100% of the time.

Reply:
I think it's important to Close the channel after we are done. Otherwise, subsequent tests will use the cached channel in the platform

Reply:
Done